src/ext/serializer.c

```diff
@@ -0,0 +1,153 @@
+    return 1;
+}
+#else
+static int write_hash_table(mpack_writer_t *writer, HashTable *ht TSRMLS_DC) /* {{{ */
+{
+    zval *tmp;
+    zend_string *string_key;
+    zend_ulong num_key;
+    int is_assoc = -1;
+
+    ZEND_HASH_FOREACH_KEY_VAL_IND(ht, num_key, string_key, tmp) {
+        if (is_assoc == -1) {
+            is_assoc = string_key != NULL ? 1 : 0;
+            if (is_assoc == 1) {
+                mpack_start_map(writer, zend_hash_num_elements(ht));
+            } else {
+                mpack_start_array(writer, zend_hash_num_elements(ht));
+            }
+        }
+        if (is_assoc == 1) {
+            mpack_write_cstr(writer, ZSTR_VAL(string_key));
+        }
+        if (msgpack_write_zval(writer, tmp TSRMLS_CC) != 1) {
+            return 0;
+        }
+    }
+    ZEND_HASH_FOREACH_END();
+
+    if (is_assoc) {
+        mpack_finish_map(writer);
+    } else {
+        mpack_finish_array(writer);
+    }
+    return 1;
+}
+#endif
+
+static int msgpack_write_zval(mpack_writer_t *writer, zval *trace TSRMLS_DC) /* {{{ */
+{
+    switch (Z_TYPE_P(trace)) {
+        case IS_ARRAY:
+            if (write_hash_table(writer, Z_ARRVAL_P(trace) TSRMLS_CC) != 1) {
+                return 0;
+            }
+            break;
+        case IS_DOUBLE:
+            mpack_write_double(writer, Z_DVAL_P(trace));
+            break;
+        case IS_LONG:
+            mpack_write_int(writer, Z_LVAL_P(trace));
+            break;
+        case IS_NULL:
+            mpack_write_nil(writer);
+            break;
+#if PHP_VERSION_ID < 70000
+        case IS_BOOL:
+            mpack_write_bool(writer, Z_BVAL_P(trace) == 1);
+            break;
+        case IS_STRING:
+            mpack_write_cstr(writer, Z_STRVAL_P(trace));
+            break;
+#else
+        case IS_TRUE:
+        case IS_FALSE:
+            mpack_write_bool(writer, Z_TYPE_P(trace) == IS_TRUE);
+            break;
+        case IS_STRING:
+            mpack_write_cstr(writer, ZSTR_VAL(Z_STR_P(trace)));
+            break;
+#endif
+        default:
+            if (DDTRACE_G(strict_mode)) {
+                zend_throw_exception_ex(spl_ce_InvalidArgumentException, 0 TSRMLS_CC,
+                                        "Serialize values must be of type array, string, int, float, bool or null");
+            }
+            return 0;
+            break;
+    }
+    return 1;
+}
+
+int ddtrace_serialize_simple_array(zval *trace, zval *retval TSRMLS_DC) {
+    // encode to memory buffer
+    char *data;
+    size_t size;
+    mpack_writer_t writer;
+    mpack_writer_init_growable(&writer, &data, &size);
+    if (msgpack_write_zval(&writer, trace TSRMLS_CC) != 1) {
+        mpack_writer_destroy(&writer);
+        free(data);
+        return 0;
+    }
+    // finish writing
+    if (mpack_writer_destroy(&writer) != mpack_ok) {
+        free(data);
+        return 0;
+    }
+#if PHP_VERSION_ID < 70000
+    ZVAL_STRINGL(retval, data, size, 1);
+#else
+    ZVAL_STRINGL(retval, data, size);
```

Comment:
This operation copies string value over. Which for a brief while doubles the memory use.
Would it be possible to create empty zend_string at the beginning of this function, and pass pointers to it char* and size ?

Reply:
Unless I misunderstood your suggestion, I don't think this is possible to do. I spent a lot of time digging for a solution and when I came up empty, I asked on Room 11 and [Niki said it wasn't possible](https://chat.stackoverflow.com/transcript/message/45874560#45874560). Or did I just misunderstand what you were suggesting? :)

Reply:
Looks like with "growable" approach its indeed not possible due to how PHP stores zend_string :( 

Thanks for looking into it!